### PR TITLE
Require TestCaseSource and ValueSource named members to be static.

### PR DESCRIPTION
**Pull request by CharliePoole**
This fiixes #698. It implements the proposal I made in #698, which has as yet not received any comments, so the review of this PR should encompass both design and code review.

Converting all our own tests to use statics for TestCaseSource and ValueSource was tedious but not difficult. The only unexpected issue is that it may require care in some cases to ensure that statics are initialized in the proper order. There was only one class where this came up, DelayedConstraintTests, and it was resolved using a class constructor.

### CODE REVIEW

File: src/NUnitFramework/framework/Attributes/TestCaseSourceAttribute.cs [hunk 1]
```diff
@@ -56,12 +56,10 @@ public TestCaseSourceAttribute(string sourceName)
         /// </summary>
         /// <param name="sourceType">The Type that will provide data</param>
         /// <param name="sourceName">The name of the method, property or field that will provide data</param>
```
Comment (rprouse) [inline src/NUnitFramework/framework/Attributes/TestCaseSourceAttribute.cs]: Should we add the fact that the sourcName must be static to the XmlDocs?

Comment (CharliePoole) [inline src/NUnitFramework/framework/Attributes/TestCaseSourceAttribute.cs]: Good idea!

File: src/NUnitFramework/framework/Attributes/TestCaseSourceAttribute.cs [hunk 2]
```diff
@@ -179,67 +184,87 @@ public IEnumerable<ITestCaseData> GetTestCasesFor(MethodInfo method)
...
         {
-            List<TestMethod> tests = new List<TestMethod>();
-
-            foreach (ParameterSet parms in GetTestCasesFor(method))
-                tests.Add(_builder.BuildTestMethod(method, suite, parms));
-
-            return tests;
+            var parms = new ParameterSet();
+            parms.RunState = RunState.NotRunnable;
+            parms.Properties.Set(PropertyNames.SkipReason, "The name specified on a TestCaseSourceAttribute must refer to a static field, property or method.");
```
Comment (rprouse) [inline src/NUnitFramework/framework/Attributes/TestCaseSourceAttribute.cs]: In both constructors, the parameter in question is `sourceName`. Should we be more specific in this message and say something like "The sourceName parameter on the TestCaseSourceAttribute..."?

File: src/NUnitFramework/framework/Attributes/ValueSourceAttribute.cs [hunk 3]
```diff
@@ -46,7 +46,7 @@ public class ValueSourceAttribute : DataAttribute, Interfaces.IParameterDataSour
         /// <param name="sourceName">The name of the data source to be used</param>
         public ValueSourceAttribute(string sourceName)
         {
-            this.sourceName = sourceName;
+            SourceName = sourceName;
```
Comment (rprouse) [inline src/NUnitFramework/framework/Attributes/ValueSourceAttribute.cs]: Same point about TestCaseSource about including the restrictions in the XmlDocs.

File: src/NUnitFramework/framework/Attributes/ValueSourceAttribute.cs [hunk 4]
```diff
@@ -98,37 +98,57 @@ public IEnumerable GetData(ParameterInfo parameter)
...
+                    ThrowInvalidDataSourceException();
+                }
             }
+
             return null;
         }
 
+        private static void ThrowInvalidDataSourceException()
+        {
+            throw new InvalidDataSourceException("The name specified on a ValueSourceAttribute must refer to a static field, property or method.");
```
Comment (rprouse) [inline src/NUnitFramework/framework/Attributes/ValueSourceAttribute.cs]: Same comment as TestCaseSource about specifying the sourceName parameter.

File: src/NUnitFramework/framework/nunit.framework-2.0.csproj [hunk 5]
```diff
@@ -198,6 +198,7 @@
     <Compile Include="Internal\Execution\WorkItemState.cs" />
     <Compile Include="Interfaces\IApplyToContext.cs" />
     <Compile Include="Interfaces\IApplyToTest.cs" />
+    <Compile Include="Internal\InvalidDataSourceException.cs" />
```
Comment (rprouse) [inline src/NUnitFramework/framework/nunit.framework-2.0.csproj]: Doesn't this also need to be added to the CF project?

File: src/NUnitFramework/tests/Attributes/TestCaseSourceTests.cs [hunk 6]
```diff
@@ -87,72 +84,28 @@ public void SourceCanBeStaticField(string source)
         static object[] StaticField =
             { new object[] { "StaticField" } };
 
-        [Test, TestCaseSource("InstanceField")]
-        public void SourceCanBeInstanceField(string source)
+        [Test]
+        public void SourceUsingInstanceFieldIsNotRunnable()
```
Comment (rprouse) [inline src/NUnitFramework/tests/Attributes/TestCaseSourceTests.cs]: I may have missed them, but should we also have the Static...IsRunnable tests for this and ValueSource?

**Comment by rprouse** [comment]
Other than my minor comments, looks good. :+1:

File: src/NUnitFramework/framework/Attributes/TestCaseSourceAttribute.cs (hunk 2 above)
Comment (CharliePoole) [inline src/NUnitFramework/framework/Attributes/TestCaseSourceAttribute.cs]: Will do

File: src/NUnitFramework/tests/Attributes/TestCaseSourceTests.cs (hunk 6 above)
Comment (CharliePoole) [inline src/NUnitFramework/tests/Attributes/TestCaseSourceTests.cs]: They are in the same files actually... in the gaps between the changed items.

File: src/NUnitFramework/framework/nunit.framework-2.0.csproj (hunk 5 above)
Comment (CharliePoole) [inline src/NUnitFramework/framework/nunit.framework-2.0.csproj]: Yes!

**Comment by CharliePoole** [comment]
I'll make those changes and merge.

File: src/NUnitFramework/tests/Attributes/TestCaseSourceTests.cs (hunk 6 above)
Comment (rprouse) [inline src/NUnitFramework/tests/Attributes/TestCaseSourceTests.cs]: I guess I should expanded the gaps in the diffs :smile:

**Comment by CharliePoole** [comment]
Looks like PR #680 broke the CF build with changes in OsPlatform.cs. I'll fix along with this.

**Pull request merged**